plenum/server/replica.py

```diff
@@ -873,16 +924,21 @@ def create3PCBatch(self, ledger_id):
             if last_ordered_ts:
                 self.last_accepted_pre_prepare_time = last_ordered_ts
 
-        reqs, invalid_indices, rejects, tm = self.consume_req_queue_for_pre_prepare(
-            ledger_id, self.viewNo,
-            pp_seq_no)
-        if len(reqs) == 0:
-            self.logger.trace('{} not creating a Pre-Prepare for view no {} '
-                              'seq no {}'.format(self, self.viewNo, pp_seq_no))
-            return
+        # DO NOT REMOVE `view_no` argument, used while replay
+        # tm = self.utc_epoch
+        tm = self.get_utc_epoch_for_preprepare(self.instId, self.viewNo,
+                                               pp_seq_no)
 
-        digest = self.batchDigest(reqs)
+        reqs, invalid_indices, rejects = [], [], []
+        if not is_empty:
+            reqs, invalid_indices, rejects = self.consume_req_queue_for_pre_prepare(
+                ledger_id, tm, self.viewNo, pp_seq_no)
+            if len(reqs) == 0:
+                self.logger.trace('{} not creating a Pre-Prepare for view no {} '
+                                  'seq no {}'.format(self, self.viewNo, pp_seq_no))
+                return
```

Comment:
I see here that batch won't be created if there are no transactions to order, unless `is_empty` is `True`, so some parameter is needed to allow creation of empty batches. However, I still believe that it's better always try to add some transactions. In this case it would be better to rename `is_empty` to something like `allow_empty`, and do here something like this:
```python
reqs, invalid_indices, rejects = self.consume_req_queue_for_pre_prepare(
    ledger_id, tm, self.viewNo, pp_seq_no)
if not allow_empty and len(reqs) == 0:
    self.logger.trace('{} not creating a Pre-Prepare for view no {} '
                                  'seq no {}'.format(self, self.viewNo, pp_seq_no))
    return
```

Reply:
We can completely remove this parameter since we already check for empty queues before calling this method.